### PR TITLE
[WIP]training_tran_sang_filter_book_author

### PR DESCRIPTION
**Pull request by tran-sang-mulodo**
## Task list
- [x] Controller
- [x] Model
- [x] View
- [x] Validate form
- [x] Ajax
- [x] Route
- [ ] Helper
 - [x] ActiveModelHelper
 - [x] AssetTagHelper
 - [x] RenderingHelper
 - [x] TextHelper
 - [x] FormHelper
 - [x] UrlHelper
- [ ] Filter
 - [x] Before action
 - [x] After action

### CODE REVIEW

File: app/models/author.rb [hunk 1]
```diff
@@ -1,7 +1,13 @@
 class Author < ApplicationRecord
+  has_many :books
   validates :name, presence: true ,uniqueness: true, length: {minimum:5 , maximum:50 }
   validates :description, presence: true
   has_attached_file :image, styles: { medium: "300x300>", thumb: "100x100>" }
   validates_attachment :image, content_type: { content_type: ["image/jpeg", "image/gif", "image/png"] },
                     size: { less_than: 4.megabyte },presence: true
+
+  PAGING = 5
+  def self.getDataAuthor(params)
```
Comment (truong-dat-mulodo) [inline app/models/author.rb]: Please use `scope` replace for `self` method

File: app/models/book.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  belongs_to :author
+  validates :name, presence: true ,uniqueness: true, length: {minimum:5 , maximum:50 }
+  validates :description, presence: true
+  has_attached_file :image, styles: { medium: "300x300>", thumb: "100x100>" }
+  validates_attachment :image, content_type: { content_type: ["image/jpeg", "image/gif", "image/png"] },
+                    size: { less_than: 4.megabyte },presence: true
+
+  PAGING = 5
+  def self.getDataBook(params)
+    books = Book.includes(:author).paginate(:page => params[:page], :per_page => Book::PAGING).order('created_at desc')
```
Comment (truong-dat-mulodo) [inline app/models/book.rb]: Please use scope replace for self method

File: app/controllers/home/authors_controller.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  before_action :authenticate_user!, except: %i[index show]
+
+  def index
+    @authors = Author.getDataAuthor(params)
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def show(params: :id); end
```
Comment (truong-dat-mulodo) [inline app/controllers/home/authors_controller.rb]: Please remove `(params: :id)`

File: app/controllers/home/authors_controller.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    respond_to do |format|
+      if @authors.save
+        format.js { redirect_to home_authors_path, notice: 'Author has been created by ajax.' }
+      else
+        format.js
+      end
+    end
+  end
+
+  def edit(params: :id); end
```
Comment (truong-dat-mulodo) [inline app/controllers/home/authors_controller.rb]: Please remove `(params: :id)`. Because don't need it

File: app/controllers/home/books_controller.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  before_action :set_book, only: %i[show edit update destroy]
+  before_action :authenticate_user!, except: %i[index show]
+  def index
+    @books = Book.getDataBook(params)
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def show(params: :id); end
```
Comment (truong-dat-mulodo) [inline app/controllers/home/books_controller.rb]: Please remove (params: :id). Because don't need it

File: app/controllers/home/books_controller.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    respond_to do |format|
+      if @books.save
+        format.js { redirect_to home_books_path, notice: 'Book has been created by ajax' }
+      else
+        format.js
+      end
+    end
+  end
+
+  def edit(params: :id); end
```
Comment (truong-dat-mulodo) [inline app/controllers/home/books_controller.rb]: Please remove `(params: :id)`. Because don't need it

File: app/controllers/home/authors_controller.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+  def show
+  end
+
+  def new
+    @author = Author.new
+  end
+
+  def create
+    @authors = Author.create(post_params)
```
Comment (truong-dat-mulodo) [inline app/controllers/home/authors_controller.rb]: Please use `new`

**Comment by truong-dat-mulodo** [comment]
@tran-hai-mulodo 
I reviewed. Please check it again. Thanks.

File: db/seeds.rb [hunk 8]
```diff
@@ -30,3 +30,21 @@
   end
 end
 
+photos = Dir.glob('app/assets/images/no-image-icon-6.png')
```
Comment (tran-hai-mulodo) [inline db/seeds.rb]: @tran-sang-mulodo 

We should create new seed file(e.g: db/seeds/authors.rb) for Authors data.
In file "db/seeds.rb", you can only use "require" it:
`require File.dirname(__FILE__) + '/seeds/authors.rb'`
or you can use "seedbank" gem to create seed data  for each environment, share seeds between environments or specify dependencies to load your seeds

File: db/seeds.rb [hunk 9]
```diff
@@ -30,3 +30,21 @@
...
+file = Rack::Test::UploadedFile.new(photos.sample, 'image/png')
+10.times do |i|
+  Author.create(name: FFaker::Name.name, image: file, description: FFaker::Book.description)
+end
+
+Author.all.each do |author|
+    author.books.create(name: FFaker::Book.title, image: file, description: FFaker::Book.description)
+end
+
```
Comment (tran-hai-mulodo) [inline db/seeds.rb]: @tran-sang-mulodo 

Please remove unuse code lines in this file. Thanks.

File: db/migrate/20181204062959_add_attachment_image_to_articles.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class AddAttachmentImageToArticles < ActiveRecord::Migration[5.1]
+  def change
+    add_attachment :articles, :image
+  end
+end
```
Comment (tran-hai-mulodo) [inline db/migrate/20181204062959_add_attachment_image_to_articles.rb]: @tran-sang-mulodo 

We should add new line at the end of file. Thanks.
